### PR TITLE
Add ORCA GUC's for dynamic index/bitmap scan

### LINKED ISSUE #16885: Providing `optimizer_enable_indexscan` and `optimizer_enable_bitmapscan` to ask GPORCA to use index or not
## Feature Request

**Is your feature request related to a problem? Please describe:**

Hi, Team.

Currently when there exists some indexes at a partition table, GPORCA tends to utilize these indexes during a SQL JOIN,
but then it is very easy to choose nested loop join if the table is joined with some other table on these indexed columns.

Usually we find the following query plan pattern happens when joining a relatively small table with a large partition table.
1. At the first stage (dynamic) bitmap index scans and a bitmap heap scan are used on the partition table
2. On the (relatively) small table, a seq scan is performed
3. A nested loop join works on data of the seq scan (small table) output and bitmap heap scan rows

And this usually causes unending execution (exceeds many hours).

However, if the partition table does not have indexes (by dropping these indexes), GPORCA can utilize dynamic seq scan then hash join processing; usually the same query can be finished within several minutes (with some higher memory utilization).

Such slowness (when indexes exist) is very significant when handling joining ~10^4 rows against partition table with lots of rows.

Usually such tweak in PGSQL is achieved by disabling query plan related GUCs, [`enable_indexscan`](https://docs.vmware.com/en/VMware-Greenplum/6/greenplum-database/ref_guide-config_params-guc-list.html#enable_indexscan), [`enable_bitmapscan`](https://docs.vmware.com/en/VMware-Greenplum/6/greenplum-database/ref_guide-config_params-guc-list.html#enable_bitmapscan), and [`enable_nestloop`](https://docs.vmware.com/en/VMware-Greenplum/6/greenplum-database/ref_guide-config_params-guc-list.html#enable_nestloop).

But I find that the above GUCs does not work for GPORCA, and there lacks equivalent GUCs on index scan, bitmap index scan, and nestloop join.

**Describe the feature you'd like to request:**

I think GPDB should provides `optimizer_enable_indexscan` and `optimizer_enable_bitmapscan` for hinting GPORCA query planner.

If possible, I think such GUCs should be included in GPDB v6, not just in GPDB7, since many users are still using GPDB v6.

Best Regards.

### PR DESCRIPTION
**Pull request by THANATOSLAVA**
Add the following GUC's

optimizer_enable_dynamicindexscan
optimizer_enable_dynamicbitmapscan

Fix #16885

`optimizer_enable_bitmapscan` used to control the switch of dynamic bitmap scan as well. This is no more the case with the addition of `optimizer_enable_dynamicbitmapscan`.

### CODE REVIEW

**Review by vraghavan78 — APPROVED** [review]

File: src/backend/gpopt/config/CConfigParamMapping.cpp [hunk 1]
```diff
@@ -380,6 +380,14 @@ CConfigParamMapping::PackConfigParamInBitset(
 		bitmap_index_bitset->Release();
 	}
 
+	// disable bitmap scan if the corresponding GUC is turned off
+	if (!optimizer_enable_dynamicbitmapscan)
+	{
+		// disable dynamic index scan if the corresponding GUC is turned off
```
Comment (dgkimura) [inline src/backend/gpopt/config/CConfigParamMapping.cpp:386]: s/disable bitmap scan/disable dynamic bitmap scan/

I think the bottom comment can be removed.

**Review by dgkimura — APPROVED** [review]

File: src/backend/gpopt/config/CConfigParamMapping.cpp [hunk 2]
```diff
@@ -380,6 +380,14 @@ CConfigParamMapping::PackConfigParamInBitset(
 		bitmap_index_bitset->Release();
 	}
 
+	// disable bitmap scan if the corresponding GUC is turned off
```
Comment (gpopt) [inline src/backend/gpopt/config/CConfigParamMapping.cpp:383]: disable dynamic bitmap scan

File: src/backend/gpopt/config/CConfigParamMapping.cpp (hunk 1 above)
Comment (THANATOSLAVA) [reply src/backend/gpopt/config/CConfigParamMapping.cpp:386]: Done

File: src/backend/gpopt/config/CConfigParamMapping.cpp (hunk 2 above)
Comment (THANATOSLAVA) [reply src/backend/gpopt/config/CConfigParamMapping.cpp:383]: Done

Comment (gpopt) [reply src/backend/gpopt/config/CConfigParamMapping.cpp:383]: nope

**Comment by THANATOSLAVA** [comment]
> We currently disable `ExfSelect2DynamicBitmapBoolOp` already with `optimizer_enable_bitmapscan`. Should we remove this as part of adding this GUC?

Yeah I thought about that and I don't have a strong opinion this way or the other. Generally I think GUC's should be exclusive. But I am also reluctant to change the existing behavior that bestowed `optimizer_enable_bitmapscan` super power of controlling dynamic bitmap scans. Lemme check with PM and see if they have any concern.

File: src/backend/gpopt/config/CConfigParamMapping.cpp (hunk 2 above)
Comment (THANATOSLAVA) [reply src/backend/gpopt/config/CConfigParamMapping.cpp:383]: Haven't check in

File: src/test/regress/sql/bfv_partition_plans.sql [hunk 3]
```diff
@@ -134,6 +135,7 @@ select * from mpp23195_t1,mpp23195_t2 where mpp23195_t1.i < mpp23195_t2.i;
 drop table if exists mpp23195_t1;
 drop table if exists mpp23195_t2;
 set optimizer_enable_bitmapscan=off;
+set optimizer_enable_dynamicbitmapscan=off;
```
Comment (chrishajas) [inline src/test/regress/sql/bfv_partition_plans.sql:138]: I'm curious why we're disabling these gucs here, we're turning off these gucs for the rest of the tests (probably unintentionally...)

**Review by chrishajas — APPROVED** [review]
Code change looks good

File: src/test/regress/sql/bfv_partition_plans.sql (hunk 3 above)
Comment (THANATOSLAVA) [reply src/test/regress/sql/bfv_partition_plans.sql:138]: It also puzzled me. I almost felt that we should be picking the dynamic bitmap scan alternative cause we have the index set up, but didn't do so afraid of changing the test's intention.

**Review by chrishajas — APPROVED** [review]

**Review by dgkimura — APPROVED** [review]

**Pull request merged**